### ISSUE TITLE
How to create EC2 instance with public IP enabled 

### ISSUE THREAD

**Issue by binchenX** [aws/aws-sdk-go-v2#1190]
Confirm by changing [ ] to [x] below:
- [x] I've gone though the [API reference](https://github.com/aws/aws-sdk-go-v2/issues)
- [x] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-sdk-go) for answers

**Describe the question**
Since the `ec2.RunInstancesInput` has not argument  to enable the public IP, so I think one (1*) way is to 1) create an interface explicitly  and 2)  set the `InstanceNetworkInterfaceSpecification.AssociatePublicIpAddress` but I met following error:

```
Got an error creating an instance:
operation error EC2: RunInstances, https response error StatusCode: 400, RequestID: 44b3be40-64cb-4ee9-804a-88302c80db88, api error InvalidParameterValue: Each network interface requires a device index.
```

Below is the code snippet used (error check and other parts are removed for brevity. )
```
	network, err := client.CreateNetworkInterface(context.TODO(), &ec2.CreateNetworkInterfaceInput{
		SubnetId: subnetId,
		Groups: []string{sgID},
         })

	input := &ec2.RunInstancesInput{
		ImageId:      aws.String(AmiUbuntu),
		InstanceType: types.InstanceTypeT2Micro,
		KeyName:      aws.String(keyPair),
		MaxCount:     1,
		MinCount:     1,
		NetworkInterfaces: []types.InstanceNetworkInterfaceSpecification{
			{
				Description:              aws.String("the primary device eth0"),
				DeviceIndex:              0,
				NetworkInterfaceId:       network.NetworkInterface.NetworkInterfaceId,
				AssociatePublicIpAddress: true,
				SubnetId:                 subnetId,
				Groups:                   []string{sgID},
			},
		},
```
Not sure what is missing here and the error message doesn't quite help. Appreciate your help!

------------------
1* According to the [doc](https://docs.aws.amazon.com/vpc/latest/userguide/vpc-ip-addressing.html), the other way is to enable the subnet attribute to let all the instance created in that subnet have an public IP automatically. I didn't do this way since 1) I don't want all the instance in that subnet to have an public IP and 2) I didn't see the option to have that attribute enabled when *creating* the subnet but I do see that attribute in the output of the subnet creating.

**Comment by varnastadeus** [comment]
@binchenX I think this is related to #1178

**Comment by jasdel** [comment]
Thanks for reaching out @binchenX. This issue is similar to https://github.com/aws/aws-sdk-go-v2/issues/1107 Where the EC2 API model is missing metadata instructing the SDK how it should serialize these operations. We're working with the EC2 service team on how best to address these operations.

**Comment by jasdel** [comment]
This issue has been resolved in #1195 in the latest release tag, `service/ec2/v1.6.0`.  You can update your application to use the latest version with

```sh
go get -d github.com/aws/aws-sdk-go-v2/service/ec2@v1.6.0
```

While this release fixes the client to be able to send all request parameters to the API, a breaking change to the generated code was needed. The parameters like `MaxCount`, `MinCount`, and  `DeviceIndex` are pointer types now, not values. You can use either addresses, or the `aws.Int32` utilities to not need to create a local variable when assigning the value.
